File: src/main/java/algo/greedy/SplitArrayIntoConsecutiveSubsequences_659.java
package algo.greedy;

import java.util.ArrayList;

public class SplitArrayIntoConsecutiveSubsequences_659 {
  public static boolean isPossible(int[] nums) {
    var len = nums.length;
    var color = 1;
    var shortSeq = new ArrayList<SeqInfo>();   // sequences whose len < 3
    var sequences = new ArrayList<SeqInfo>();  // sequences whose len >= 3
    shortSeq.add(new SeqInfo(1, 1, 0));
    for (var i = 1; i < len; i++) {
      if (nums[i] >= nums[i - 1] + 2) {
        color++;
        shortSeq.add(new SeqInfo(color, 1, i));
      } else {
        var found = false;

        // add to seq whose len < 3
        if (!shortSeq.isEmpty()) {
          for (var j = 0; j < shortSeq.size(); j++) {
            var candidate = shortSeq.get(j);
            if (nums[candidate.end] == nums[i] - 1) {
              found = true;
              candidate.len++;
              candidate.end = i;
              if (candidate.len == 3) {
                shortSeq.remove(j);
                sequences.add(candidate);
              }
              break;
            }
          }
        }

        // add to sequences whose len >= 3
        if (!found && !sequences.isEmpty()) {
          for (var candidate : sequences) {
            if (nums[candidate.end] == nums[i] - 1) {
              found = true;
              candidate.len++;
              candidate.end = i;
              break;
            }
          }
        }

        // create a new sequence
        if (!found) {
          color++;
          shortSeq.add(new SeqInfo(color, 1, i));
        }
      }
    }
    return shortSeq.isEmpty();
  }

  static class SeqInfo {
    public int color;
    public int len;
    public int end;
    public SeqInfo(int color, int len, int end) {
      this.color = color;
      this.len = len;
      this.end = end;
    }
  }

  public static void main(String[] args) {
    System.out.println(isPossible(new int[]{1,2,3,3,4,5}));
    System.out.println(isPossible(new int[]{1,2,3,3,4,4,5,5}));
    System.out.println(isPossible(new int[]{1,2,3,4,4,5}));
    System.out.println(isPossible(new int[]{1,2,3,5,5,6,7}));
    System.out.println(isPossible(new int[]{4,5,6,7,7,8,8,9,10,11}));
    System.out.println(isPossible(new int[]{10,11,11,12,12,12,13,13,13,13,14}));
    System.out.println(isPossible(new int[]{1,2,3,4,5,5,6,7}));
  }
}

File: src/main/kotlin/algo/greedy/SplitArrayIntoConsecutiveSubsequences_659.kt
package algo.greedy

class KtSplitArrayIntoConsecutiveSubsequences_659 {
    fun isPossible(nums: IntArray): Boolean {
        val shortSeq: MutableList<Seq> = mutableListOf()
        val seqs: MutableList<Seq> = mutableListOf()
        nums.indices.forEach { i ->
            val num = nums[i]
            if (i == 0 || nums[i] - nums[i - 1] > 1) {
                shortSeq.add(Seq(num, 1))
            } else {
                // try to add to sequences whose len < 3
                var added = false
                run shortLoop@ {
                    shortSeq.indices.forEach {
                        val seq = shortSeq[it]
                        if (seq.end == num - 1) {
                            added = true
                            seq.end = num
                            seq.len++
                            if (seq.len == 3) {
                                shortSeq.removeAt(it)
                                seqs.add(seq)
                            }
                            return@shortLoop
                        }
                    }
                }
                // try to add to sequences whose len >= 3
                if (!added) {
                    run longLoop@ {
                        seqs.forEach { seq ->
                            if (seq.end == num - 1) {
                                added = true
                                seq.end = num
                                seq.len++
                                return@longLoop
                            }
                        }
                    }
                }
                if (!added) {
                    shortSeq.add(Seq(num, 1))
                }
            }
        }
        return shortSeq.isEmpty()
    }
}

data class Seq(var end: Int, var len: Int)

fun main() {
    val split = KtSplitArrayIntoConsecutiveSubsequences_659()
    println(split.isPossible(intArrayOf(1,2,3,3,4,5)))
    println(split.isPossible(intArrayOf(1,2,3,3,4,4,5,5)))
    println(split.isPossible(intArrayOf(1,2,3,4,4,5)))
    println(split.isPossible(intArrayOf(1,2,3,5,5,6,7)))
}